proposals/0165-async-vote-execution.md

```diff
@@ -0,0 +1,218 @@
+transactions in a block without checking fee payers. The exact calculation
+algorithm is listed in next section.
+* `Replay Tip Bankhash`: The bankhash as we know it today. It is calculated
+after executing all transactions in a block, checking fee payers for all.
+
+## Detailed Design
+
+### Allow leader to skip execution of transactions (Bankless Leader)
+
+There is already on-going effort to totally skip execution of all transactions
+when leader pack new blocks. See SIMD 82, SIMD 83, and related trackers:
+https://github.com/anza-xyz/agave/issues/2502
+
+Theoretically we could reap some benefit without Bankless Leader, the leader
+packs as normal, while other validators only replay votes first, then later
+execute other transactions and compare with the bankhash of the leader. But in
+such a setup we gain smaller speedup without much benefits, it is a possible
+route during rollouts though.
+
+### Calculate vote only hash executing votes only and vote on selected forks
+
+Two new fields will be added to `TowerSync` vote transaction:
+
+* `replay_tip_hash`: This is the hash as we know it today.
+* `replay_tip_slot`: This is the slot where the replay tip hash is calculated.
+
+The `hash` and `slot` in the `TowerSync` transaction will be updated to
+the vote only hash. The vote only hash is calculated as follows:
+
+1. Sort all vote accounts with non-zero stake in the current epoch by
+vote account pubkey.
+
+2. Calculate vote account hash by hashing (vote account pubkey,
+serialized vote state) in the order given.
+
+3. Calculate vote only hash by hashing:
+
+* vote only hash of the parent bank
+* vote account hash calculated above
+* block-id of the current bank
+
+This step is optimistic in the sense that validators do not check the fee
+payers when executing the vote transactions in a block. They assume vote
+transactions will not fail due to insufficient fees, apply the execution
+results to select the correct fork, then immediately vote on the bank with
+only the hash result of the vote transactions.
+
+This is safe because the list of validators and their corresponding stake
+uses the leader scheduler epoch stakes, which is calculated at the beginning
+of last Epoch. Because full execution is never behind the optimistic execution
+by more than one Epoch, the epoch stakes used is stable and correct.
+
+To make sure the vote casted would be the same as that after replaying the
+whole block, we need to be consistent on whether we mark the block dead, so
+that the ephemeral hash vote doesn't vote on a block which will be marked
+dead later. Currently a block can be dead for the following reasons:
+
+1. Unable to load data from blockstore
+2. Invalid block (wrong number of ticks, duplicate block, bad last fec, etc)
+3. Error while set root
+4. Invalid transaction
+
+For the first two, the same check can be performed computing ephemeral hash.
+We will set root on a bank only when it has full hash computed later, so the
```

Comment:
what does "set root" mean in this context?

Reply:
This is SetRootError in BlockstoreProcessorError. I think we will keep this check in the new world.

Reply:
it doesn't mark the block dead though, just kills the replay thread.

Reply:
Hmm that's true, removed this check